[16.0][IMP+FIX] spreadsheet*_oca: New field to name the pivot table and fixed error adding to new sheet of dashboard

CarlosRoca13 opened the pull request:
cc @Tecnaiva TT50012 and TT50206

ping @pedrobaeza @chienandalu

CarlosRoca13 committed:
[IMP] spreadsheet*_oca: Add field to define name of table and restructure order of fields

CarlosRoca13 committed:
[FIX] spreadsheet_dashboard_oca: Error with dashboard_sheet selected

The error is happening because the dashboard where
the data has to be added is not being showed.

With this changes the field to define the
dashboard is required and is showed.

Added a way to limit the search to the dashboards
that can be edited

CarlosRoca13 committed:
[IMP] spreadsheet_oca: Allow generate dynamic tables

On dashboards exists tables that are updated dynamically,
with this changes we have the option to generate this
dynamic tables using the pivot inserter.

Added a button to duplicate and change to dynamic
table the selected table on spreadsheet editor.

If a table with multiple row goups generated, the
way to generate the dynamic table will be set the
number of rows for each level of indentation.

Example:
2 rowGroupBys
3 number of rows
- val1
    - subval1.1
    - subval1.2
    - subval1.3
- val2
    - subval2.1
    - subval2.2
    - subval2.3
- val3
    - subval3.1
    - subval3.2
    - subval3.3

CarlosRoca13 committed:
[IMP] spreadsheet_oca: Allow download spreadsheet as XLSX

CarlosRoca13 commented:
Added improvements to allow generate dynamic tables as showed in next gif and added an option to download the spreadsheet as xlsx (this option is available inside a spreadsheet on menu item `File`)
![dynamic_table_example](https://github.com/user-attachments/assets/cb389a50-5f0e-4d30-b6f4-2da81db2b474)

carolinafernandez-tecnativa commented:
Thanks for the improvements :) Some comments:

1- I have checked and when downloading XLS is not respeting spreadsheet format.

Example:

<img width="790" alt="image" src="https://github.com/user-attachments/assets/f10c59ff-7f64-4c92-af87-beaddb752e62">

<img width="900" alt="image" src="https://github.com/user-attachments/assets/f8f1cca8-e964-4cf5-b476-99cdde0539f7">

2- Dynamic field i think it must be visible only when trying to generate spreadhseet from pivot view, as if we try to create in tree view form is not working.

<img width="1427" alt="image" src="https://github.com/user-attachments/assets/c4a46bf0-b1f2-4dc7-90f9-0c28b53afda0">

3- Database source is not filling automatically when creating spreadhseet from tree view. What should i add in this field?

![image](https://github.com/user-attachments/assets/9b38f82b-7cd8-43d5-b11f-cf15a73ce4da)

![image](https://github.com/user-attachments/assets/ae716bb2-2064-4a9c-9fe9-fb79ffb21275)

CarlosRoca13 commented:
> 1- I have checked and when downloading XLS is not respeting spreadsheet format.

The XLSX data is generated by the core module. If you check the behavior in the odoo enterprise module, you will see that the result is the same.

For other comments I will do the changes, thanks :)

CarlosRoca13 committed:
[FIX] spreadsheet_oca: Make last changes compatibles with tree and graph

Since this changes the field datasource_name does not work with
tree and graph. Now the datasource_name is set as data name for the
created data.

Furthermore, the capability to create dynamic tables has no sense on
tree and graph, so it has been limited for pivots.

CarlosRoca13 commented:
@carolinafernandez-tecnativa changes done :smile: 

The value of Datasource Name is the name that will be filled to identify the table on the spreadsheet.

chienandalu commented on spreadsheet_dashboard_oca/models/spreadsheet_dashboard.py:
```diff
@@ -42,3 +43,24 @@ def _compute_can_edit(self):
+            raise NotImplementedError(_("Search operation not supported"))
+        if not isinstance(value, bool):
+            raise ValidationError(_("The value has to be a boolean"))
+        no_edit_ids = (
+            self.env["ir.model.data"]
+            .search(
+                [
+                    ("model", "=", self._name),
+                    ("module", "!=", "__export__"),
+                    ("noupdate", "=", 0),
+                ]
+            )
+            .mapped("res_id")
+        )
```
I created a brand new dashboard but I can't make it to show up in the wizard. Am I doing something wrong

CarlosRoca13 replied:
You have to add a spreadsheet to the panel to be able to show it

chienandalu replied:
I mean, I can't select any dashboard in the wizard

CarlosRoca13 replied:
![example](https://github.com/user-attachments/assets/c3c51ead-6cb6-4a48-b7c1-b5d2c673c110)

I am able to select it.

chienandalu replied:
I was trying to dashboard new sheet. I was expecting to let me choose a dashboard to wich apend a sheet. Did I understand it wrong?

CarlosRoca13 replied:
Yes :sweat_smile: that option is to add a sheet to a created spreadsheet for the dashboard

Example:
![new sheet example](https://github.com/user-attachments/assets/d0d040cb-5f2d-4452-8bca-444abf4dc58b)

This option can be used to set data that has to be hidden but it has to be used to make some operations

CarlosRoca13 replied:
I think that option that you say can be set, I will see how to add it 

Thanks :smile:

chienandalu replied:
Ah, ok. See? I was getting it wrong :sweat_smile:

CarlosRoca13 replied:
Added new option to create new dashboards :smile:

CarlosRoca13 committed:
[IMP] spreadsheet_dashboard_oca: Add option to create new dashboard

CarlosRoca13 commented:
The tables imported from the tree are dynamic by nature. So I have made some changes so that the dynamic selector is selected by default and that the value of the number of rows is filled with the value that was used in the threshold

Example:
![tables](https://github.com/user-attachments/assets/38779315-3b58-4f0b-8b5b-18c5808d353e)

pedrobaeza reviewed (approved):
Tested and working.

Thanks for all the fixes and the new excellent features!

AntoniRomera reviewed (approved):
LGTM
We have in mind to start the migration to 17.0, if you merge this PR we could use it as a starting point.

carolinafernandez-tecnativa reviewed (approved):
Functional review LGTM :) Thanks!

pedrobaeza commented:
/ocabot merge minor
